features/251-templates/index.md

```diff
@@ -0,0 +1,177 @@
+This document introduces a new way to organize templates with localization support.
+
+
+## Template
+
+The template is represented by the following struct
+
+```golang
+type TemplateConfiguration struct {
+  Type        string
+  LanguageTag string
+  Key         string
+  // The following fields are not user-facing.
+  // URI         string
+  // ContentMD5  string
+}
+```
+
+### Template Type
+
+Each template must have a type. The type identifies how the template is used by Skygear.
+
+The following listing is the initially defined types.
+
+```golang
+type TemplateConfigurationType string
+const (
+  TemplateConfigurationTypeForgotPasswordEmailTXT             TemplateConfigurationType = "forgot_password_email.txt"
+  TemplateConfigurationTypeForgotPasswordEmailHTML            TemplateConfigurationType = "forgot_password_email.html"
+  TemplateConfigurationTypeForgotPasswordResetHTML            TemplateConfigurationType = "forgot_password_reset.html"            // Keep URL option
+  TemplateConfigurationTypeForgotPasswordSuccessHTML          TemplateConfigurationType = "forgot_password_success.html"          // Keep URL option
+  TemplateConfigurationTypeForgotPasswordErrorHTML            TemplateConfigurationType = "forgot_password_error.html"            // Keep URL option
+  TemplateConfigurationTypeWelcomeEmailTXT                    TemplateConfigurationType = "welcome_email.txt"
+  TemplateConfigurationTypeWelcomeEmailHTML                   TemplateConfigurationType = "welcome_email.html"
+  TemplateConfigurationTypeUserVerificationGeneralErrorHTML   TemplateConfigurationType = "user_verification_general_error.html"  // Keep URL option
+  TemplateConfigurationTypeUserVerificationMessageTXT         TemplateConfigurationType = "user_verification_message.txt"
+  TemplateConfigurationTypeUserVerificationMessageHTML        TemplateConfigurationType = "user_verification_message.html"
+  TemplateConfigurationTypeUserVerificationSuccessHTML        TemplateConfigurationType = "user_verification_success.html"        // Keep URL option
+  TemplateConfigurationTypeUserVerificationErrorHTML          TemplateConfigurationType = "user_verification_error.html"          // Keep URL option
+)
+```
+
+### Template Language Tag
+
+Each template may optionally have a language tag. In the future, we may support sending localized message. The language tag is specified in [BCP47](https://tools.ietf.org/html/bcp47).
+
+### Template Key
+
+If the developer has configured `email` and `secondary_email` as login IDs. User Verification supports verifying both with different templates. Therefore, `key` is required to differentiate different instances of the same type of the template.
+
+## The `templates` directory next to `skygear.yaml`
+
+The developer organizes their templates in the `templates` directory next to `skygear.yaml`.
+
+The templates must be organized in the following way in order to be recognized.
+
+```
+templates[/<language-tag>][/<key>]/<type>
```

Comment:
An ambiguous edge case would be `templates/en-US/welcome_email.txt`. So if there is just three segments in the path, the middle segment would be recognized as language tag if valid, otherwise a key?

Reply:
Actually the `type` determines whether `key` is required or never required. These is no case that `key` is optional. Since `welcome_email` never require `key`, the segment must be language tag. Your concern is problem if there is a template whose `key` is optional.